flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/CatalogManager.java

```diff
@@ -874,7 +876,24 @@ public ResolvedCatalogTable resolveCatalogTable(CatalogTable table) {
         if (table instanceof ResolvedCatalogTable) {
             return (ResolvedCatalogTable) table;
         }
+
         final ResolvedSchema resolvedSchema = table.getUnresolvedSchema().resolve(schemaResolver);
+
+        final List<String> physicalColumns =
+                DataType.getFieldNames(resolvedSchema.toPhysicalRowDataType());
```

Comment:
Just `ResolvedSchema#getColumnNames` more than `toPhysicalRowDataType()`? You don't need to build a data type to destructure it in the same statement. We also don't need to filter out metadata columns, because we can't really have columns with duplicate names where one of them is a partition column (because partition columns are matched all around the code with names, including `table.getPartitionKeys()` itself.

Reply:
This check is about semantics. It should not be allowed to define partition columns on metadata column. This would mess up the handling in the sources. This behavior is similar to primary keys which are also only supported on physical columns.

Reply:
I hope I could address your feedback by accessing the schema columns directly.